### ISSUE TITLE
Add the ability to track character cards and enhancements

### ISSUE THREAD

**Issue by joewesch** [Lurkars/gloomhavensecretariat#640]
### Is your feature request related to a problem? Please describe.

When a character levels up, they choose a new card to unlock from a set of their locked cards. I sometimes run multiple campaigns with different sets of people, so when I switch between campaigns I need to reset the physical cards for the characters that are being played. Also, there are times where people store their cards incorrectly and just don't remember which cards they picked for each level.

In addition, different people may select different enhancements to their cards, so I need to be able to reset a card back to normal or re-apply an enhancement someone had purchased.

### Describe the solution you'd like

It would be nice if I could track which level 2/3/4/etc card a character picked when they leveled up so when I am resetting the game to their campaign I can put it back the way it was.

In addition, it would also be nice to keep track of which enhancement (+1, poison, wound, blessing, etc.) someone applied to each of their cards.

### Describe alternatives you've considered

The simplest way to track this is obviously the Notes section of the character sheet, but that can get messy and cumbersome without some organization.

### Additional context

I used to use an app called Gloomhaven Campaign Tracker that had this ability, but I don't really want to have to keep a separate app just to keep track of these items. Here are some screenshots of that app for reference.

# Ability List
![Screenshot 2024-12-22 12 31 48 PM](https://github.com/user-attachments/assets/d831386c-a23d-4a58-839d-b46c8bfb54df)

# Choosing a new card at level 2
![Screenshot 2024-12-22 12 34 37 PM](https://github.com/user-attachments/assets/6561e5fb-e5c2-4816-bddc-337acee3ff4a)

# Enhancement screen
![Screenshot 2024-12-22 12 32 38 PM](https://github.com/user-attachments/assets/e6cf2525-59fa-483c-8cc7-3310137ac0c6)

**Comment by Lurkars** [comment]
Enhancements will be a long way off as all the character ability data is missing (except for Brute and Mindthief). 

At least I added name, card number and initiative for each character, I used OCR to get the data from the images, still a lot of work and there are still many errors and typos I guess. Anyway, in [v0.109.0](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.109.0) you can now go to the character sheet and click on the cards icon next to the character to open the abilities dialogue. You can select cards for each level and it will be saved. So basic deck management is there. I am not very enthusiastic about the UX myself, so feedback and ideas for improvement are very welcome.

**Comment by joewesch** [comment]
It's a decent start, thanks.

As far as feedback:
- Maybe a checkbox to display/hide the cards you didn't pick?
- I like the white/green/red hand icons. Maybe make them a little bigger or brighter so they are more obvious?
- The only thing I didn't really like is when you are selecting a card the higher level cards are added to the right side instead of the left. Chances are most of the time I'm going to pick one of the new highest level cards that I unlocked before picking one of the previous ones, so having to scroll down past cards 2-8 to find the level 9 cards (for example) seems backwards. Maybe they should be sorted highest to lowest on the card selection section?

**Comment by Lurkars** [comment]
Thanks, that's some good feedback!

**Comment by Lurkars** [comment]
So addressed your feedback in [v0.109.3](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.109.3). Just did a more logical change: List now only show your cards in deck. If you use the toggle to show all cards, then the hand icons as indicators are added.

Do you have more feedback here? Thanks in advance.

**Comment by joewesch** [comment]
So far I'm liking the way it works, thanks.

Let me know if we can help with any data entry. I would just need some guidance on where the existing data resides and I can just replicate the format and extend it with any missing cards that I own.

**Comment by Lurkars** [comment]
So I started with Enhancement feature. This is far from done and also will heavily depend on full ability cards data. So if you're really motivated, here are the current options:

- take the already done [Deck of the Brute](https://github.com/Lurkars/gloomhavensecretariat/blob/main/data/gh/character/deck/brute.json) as Template and create corresponding decks as plain JSON files.
- use the definitely not Frosthaven ready [Deck Editor](https://gloomhaven-secretariat.de/editor/deck?deck=brute) (also as example with the brute deck) and create ability deck there. This generates valid Deck JSON files.

**Comment by joewesch** [comment]
Very cool! I can definitely help in my free time.

**Comment by Lurkars** [comment]
So while still a bit useless without all the cards data, the first draft for Enhancement is ready in [v0.112.0](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.112.0).

Can now only be tested with Brute and Mindthief, but happy about feedback.

**Comment by Lurkars** [comment]
Some small improvements in [v0.112.1](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.112.1):
- added 4 enhancement dots for the "work in progress" abilities, so you can still set enhancement on that before real data is available
- the enhancement on retired characters can still be edited now to migrate for already retired characters

**Comment by Lurkars** [comment]
So from your feedback, reworked now the "work in progress" enhancements in [v0.112.3](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.112.3):
- only one slot visible, to add more (because other dot of card must be enhanced) double click on a dot
- if a *wip* enhancement get replaced with actual card data, the enhancement is moved to the corresponding slot, of course this is a bit tricky and depending of the logical order of the actual dots then, but order is just top down from every ability

Example tested with Tinkerers **Flamethrower**:
from:
![Image](https://github.com/user-attachments/assets/3ca63c11-2594-4a28-a1a3-0b5c786d26e3)
migrated to:
![Image](https://github.com/user-attachments/assets/7d632c3f-4599-480a-8a35-4bb34273a808)
without enhancements looks like this
![Image](https://github.com/user-attachments/assets/6a42efe4-ddaf-43ea-9c2f-fe6f3cd76eee)

**Comment by joewesch** [comment]
@Lurkars Can you look at this dropdown list? It is constantly refreshing and I can't open it and select a new item unless I can do it in less than a second. As a workaround I've just been changing the query param in the URL to get to where I need.

![Image](https://github.com/user-attachments/assets/12ce9389-bf1f-4f11-93d8-d40f9ccc0173)

**Comment by Lurkars** [comment]
> [@Lurkars](https://github.com/Lurkars) Can you look at this dropdown list? It is constantly refreshing and I can't open it and select a new item unless I can do it in less than a second. As a workaround I've just been changing the query param in the URL to get to where I need.

Should be fixed in [v0.113.4](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.113.4).

**Comment by joewesch** [comment]
@Lurkars, maybe I'm missing something, but how do I add `xp` and `bottomXp` in the deck editor? I'm on the `fh` edition of the deck editor but I don't see them. They only show up when I add them to the JSON on the right:

<img width="529" height="329" alt="Image" src="https://github.com/user-attachments/assets/86f54e01-f4c0-4c4f-af5a-919ab4738c45" />

<img width="173" height="121" alt="Image" src="https://github.com/user-attachments/assets/6e7b813e-e3a5-4a90-b2a7-d725ba61e139" />

<img width="766" height="511" alt="Image" src="https://github.com/user-attachments/assets/607aa296-edd9-4477-873a-7f54fbed03b5" />

**Comment by Lurkars** [comment]
As said, you need to enable FH-Style in Settings for it. Anyways, there were other issues in editor. Those should be fixed now in [v0.161.0](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.161.0) and I also just added a toggle on top for FH Style, so it should be clear now:

<img width="536" height="337" alt="Image" src="https://github.com/user-attachments/assets/4cb7eccf-a30a-4b26-a2e6-438f260cc2e4" />

**Comment by joewesch** [comment]
OH! Sorry, I missed that you meant it was in Settings. Thanks, I'll take another look tonight.

**Comment by joewesch** [comment]
How do I add the forced fast or forced slow sections of Blinkblade's cards? Or Geminate's Form shift?

<img width="2681" height="2111" alt="Image" src="https://github.com/user-attachments/assets/0607e7b1-ee4d-4964-ada6-6cd0cb5d25da" />

<img width="1573" height="1126" alt="Image" src="https://github.com/user-attachments/assets/7b9706f5-bed2-41d0-94dc-72437eefe731" />

**Comment by Lurkars** [comment]
The blinkblade stuff will take a while. For Geminate, I added custom actions with `%game.action.custom.fh-switch-range%` and `%game.action.custom.fh-switch-melee%`
